woocommerce/woocommerce
# Separate product properties into property groups

### Changes proposed in this Pull Request:

This PR refactors the product implementation in the `api` package to prepare for supporting multiple product types. In the current implementation the simple product is effectively an alias of the abstract product.

#### Model changes

- The abstract product is broken out into:
  - Common - properties common to all product types in core
  - Cross Sells property
  - Delivery properties
  - Inventory properties
  - Sales tax properties
  - Shipping properties
  - Upsells property
- Interfaces were added for each property group because Typescript only allows extending a single class but supports implementing multiple interfaces. The interfaces also ensure that any product type implementing the property groups will have the correct properties for that group.
- Shared types are broken out into
  - Classes
  - Enumerations
  - Types
- The simple product class implements all of the property groups above

#### Transformer changes

The original transformer was updated to transform the common properties. A new transformer was added for each property groups.

Closes #28985 .

### How to test the changes in this Pull Request:

1. This is a significant update to the code base and should have a thorough review of the code changes.
2. Verify E2E tests run locally & in CI
3. Edit `tests/e2e/config/jest.setup.js` and comment out the product delete call (line 73)
```
//	await trashExistingProducts();
```
4. Import the sample products
5. Create a test script `tests/e2e/specs/pr-29118.js`:
```
import { HTTPClientFactory, SimpleProduct } from '@woocommerce/api';
const config = require( 'config' );

describe( 'List Products', () => {
	it( 'can list', async () => {
		const apiUrl = config.get( 'url' );
		const adminUsername = config.get( 'users.admin.username' );
		const adminPassword = config.get( 'users.admin.password' );
		const withIndexPermalinks = HTTPClientFactory.build( apiUrl )
			.withBasicAuth( adminUsername, adminPassword )
			.withIndexPermalinks()
			.create();
		const repository = SimpleProduct.restRepository( withIndexPermalinks );
		const products = await repository.read({product-id});
		const test = { name: 'testing 29118' };

		expect( products ).toEqual( expect.objectContaining( test ) );
	});
});
```
6. Run the test `npx wc-e2e test:e2e tests/e2e/specs/pr-29118.js`
7. The test will fail and dump the product json from the api. Compare the dumped data to the product in the dashboard.
8. Repeat the test with any simple product.

### Other information:




### Changelog entry

N/A
